### PR TITLE
Persistor for historical Forex Quote data

### PR DESCRIPTION
**Pull request by merlinND**

### CODE REVIEW

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  }
+}
+
+object CSVParser extends RegexParsers with java.io.Serializable {
+  def csvcombo: Parser[Quote] = (
+    datestamp~timestamp~";"~floatingpoint~";0"~datestamp~timestamp~";"~floatingpoint~";0" ^^
+    { case date~_~time~_~bid~_~date2~_~time2~_~ask => Quote(0, date, time, whatC, withC, bid, ask) }
+  )
+  
+  val datestamp: Parser[String] = "\d{9}".r
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala]: In Scala, use triple quotes if you need to escape characters:

``` scala
val datestamp: Parser[String] = """\d{9}""".r
```

File: ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  
+  // Open DB session and query handler on the QUOTES table
+  val db = Database.forURL("jdbc:sqlite:" + dbFilename + ".db", driver = "org.sqlite.JDBC")
+  lazy val QuotesTable = TableQuery[Quotes]
+  def init() = db.withDynSession { if (MTable.getTables("QUOTES").list.isEmpty) { QuotesTable.ddl.create } }
+  
+  
+  override def loadBatch(startTime: Long, endTime: Long): List[Quote] = db.withDynSession {
+    val res = QuotesTable.filter(e => e.timestamp >= startTime && e.timestamp <= endTime).invoker
+    res.list.map( r => Quote(1, r._2, Currency.fromString(r._3), Currency.fromString(r._4), r._5, r._6))
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala]: `Quote(1, ...)`
Does this `1` will be replaced by the auto-incrementing `QUOTE_ID`?

**Comment by merlinND** [comment]
If I understand correctly, what's missing now is respecting the timestamp from the historical data rather than using a fixed delay, right?

**Comment by merlinND** [comment]
Split this PR into two:
- Historical Forex data fetcher (which will be ready first)
- Another PR to use the Persistor

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 3]
```diff
@@ -63,7 +64,7 @@ class HistDataCSVFetcher(dataDir: String, currencyPair: String,
    */
   val allQuotes: List[Quote] = monthsBetweenStartAndEnd
                                 .flatMap(m => parse(bidPref + m + ".csv", askPref + m + ".csv"))
-                                .sortBy(q => q.timestamp)
+                                .sortBy(q => q.timestamp);
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala]: No `;`

File: ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  override def save(newQuotes: List[Quote]): Unit = db.withDynSession {
+    QuotesTable ++= newQuotes.map(q => (1, q.timestamp, q.whatC.toString, q.withC.toString, q.bid, q.ask))
+  }
+  
+  //TODO
+  def loadSingle(id: Int): ch.epfl.ts.data.Quote = { Quote(1,1,Currency.EUR,Currency.CHF,1,1) }
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala]: Replace dummy implementation by `???`

File: ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+  
+  //TODO
+  def save(t: ch.epfl.ts.data.Quote): Unit = { }
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala]: Replace dummy implementation by `???`

File: ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+  db.withDynSession { if (MTable.getTables("QUOTES").list.isEmpty) { QuotesTable.ddl.create } }
+  
+  
+  override def loadBatch(startTime: Long, endTime: Long): List[Quote] = db.withDynSession {
+    val res = QuotesTable.filter(e => e.timestamp >= startTime && e.timestamp <= endTime).invoker
+    res.list.map( r => Quote(1, r._2, Currency.fromString(r._3), Currency.fromString(r._4), r._5, r._6))
+  }
+
+  override def save(newQuotes: List[Quote]): Unit = db.withDynSession {
+    QuotesTable ++= newQuotes.map(q => (1, q.timestamp, q.whatC.toString, q.withC.toString, q.bid, q.ask))
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala]: `Quote(1, ...)`
Will this `1` be replaced by the auto-incrementing `QUOTE_ID`?

**Comment by merlinND** [comment]
Parsing 5 years of a single currency is too slow.

File: report/histdataPerformance.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
+On the performance of the Historical Data CSV Fetcher
+-----------------------------------------------------
+The size of our data for one currency pair (e.g. EURCHF) is about 75M records or 3.2GB (1 record is about 43 Bytes when saved in an SQLite DB). It takes an average of 10us per record to put it into the DB on my system, i.e. about 750s = 12.5min for one currency pair. Let's make that 15min to account for uncertainty in estimations. That's about 4.3 MB/s of reading, parsing and writing to the DB. So for four currencies it will take about 1h to load them into a persistor with the current system.
```
Comment (merlinND) [inline report/histdataPerformance.md]: "four currencies" -> "four currency pairs"?

File: report/histdataPerformance.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+
+When using the fetcher as an actor and reading data directly from CSV for show or simulation purposes, it will start sending the first quotes as soon as they have been read. The fetcher keeps reading from disk lazily (i.e. only when needed) and only loads into memory what's necessary.
+
+To optimize performance in a distributed system one should probably use a DB as basis (not CSV, which needs to be parsed first). Also, one could distribute DBs for different currencies on different nodes, depending on the configuration and the network traffic this could be beneficial.
```
Comment (merlinND) [inline report/histdataPerformance.md]: Nice ideas

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 9]
```diff
@@ -59,16 +60,28 @@ class HistDataCSVFetcher(dataDir: String, currencyPair: String,
...
+    }
+  }
   
   /**
-   * Index of the next quote to be fetched, incremented whenever a quote has been fetched
+   * Index of the next month to be fetched, incremented whenever a month has been fetched.
+   * And the index for the quotes in that month, updated whenever a quote was sent.
    */
-  var quoteIndex = 0
+  var currQ = allQuotes.next()
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala]: Avoid abbreviations

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 10]
```diff
@@ -59,16 +60,28 @@ class HistDataCSVFetcher(dataDir: String, currencyPair: String,
...
+  var nextQ = allQuotes.next()
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala]: Avoid abbreviations

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 11]
```diff
@@ -79,25 +92,45 @@ class HistDataCSVFetcher(dataDir: String, currencyPair: String,
...
+   * in an sqlite database of a given name.
+   * 
+   * @param   filename    Name of the DB file the quotes will be saved to, the final file
+   *                      will be called <filename>.db
+   */
   def loadInPersistor(filename: String) {
-    throw new UnsupportedOperationException("loadInPersistor is not yet implemented.");
+    val persistor = new QuotePersistor(filename)
+    while(allQuotes.hasNext) {
+      println("Loading 100000 records into persistor...")
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala]: Explicit that we are not stopping at 100000, just that it is the max size of a batch.

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 12]
```diff
@@ -79,25 +92,45 @@ class HistDataCSVFetcher(dataDir: String, currencyPair: String,
...
+      var quoteBatch = List[Quote]()
+      var batchsize = 0
+      while(batchsize < 100000 && allQuotes.hasNext) {
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala]: Refactor the number to a `val` (to have consistency between the println and the actual loop).

File: ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala [hunk 13]
```diff
@@ -79,25 +92,45 @@ class HistDataCSVFetcher(dataDir: String, currencyPair: String,
...
       }
     }
+    
   }
   
-  // TODO
+  /**
+   * Saves the quotes this fetcher has fetched (i.e. quotes that are stored in val allQuotes)
+   * in an sqlite database of a given name.
+   *
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala:117]: Explicit what happens if we want to save data to a DB and also use our fetcher afterwards. Does the fetcher rewind all the way to the start after having saved the DB?

File: ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+  override def save(newQuotes: List[Quote]): Unit = db.withDynSession {
+    // The first field of the quote (QUOTE_ID) is set to -1 but this will be 
+    // ignored and auto incremented by jdbc:sqlite in the actual DB table.
+    QuotesTable ++= newQuotes.map(q => (-1, q.timestamp, q.whatC.toString, q.withC.toString, q.bid, q.ask))
+  }
+  
+  //TODO
+  def loadSingle(id: Int): ch.epfl.ts.data.Quote = ???
+  
+  def save(q: ch.epfl.ts.data.Quote): Unit = db.withDynSession {
```
Comment (merlinND) [inline ts/src/main/scala/ch/epfl/ts/component/persist/QuotePersistor.scala]: Implement in terms of `save(List[Quote])` to easily avoid code duplication.
Hint: `save(List(q))`

**Comment by Denwid** [comment]
I implemented your improvement suggestions @merlinND .

**Comment by merlinND** [comment]
:shipit:

**Pull request merged**